### ISSUE TITLE
vimtex hangs forever in class file

### ISSUE THREAD

**Issue by oliviertilmans** [lervag/vimtex#609]
### Explain the issue

Forking issue from #598 as my perf profile hints at another bug (and could also be the one in #606 ?)

When editing class files, vimtex cause neovim (git) and vim (8.0.46) to hang forever when inserting a new line (via `o` or by appending then pressing enter)

1. Steps to reproduce
Using the MWE/vimrc provided below:
  ```bash
vim -u test.vimrc test.cls
[press Go]
```

2. Expected behaviour
A new line is inserted at the end of the file

3. Observed behaviour
(n)vim hangs forever until ^C (and then inserts the new line)

### Minimal working example

test.cls

```tex
% vim:ft=tex
\endinput%
```

### Minimal vimrc file

test.vimrc

```vim
set nocompatible

" Load Vimtex
let &rtp  = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'

filetype plugin indent on
syntax enable
```

### Perf profile

Hints towards https://github.com/lervag/vimtex/blob/master/indent/tex.vim#L132 as the source of the hangouts (called once, and apparently loops forever until ^C)

```
[...]
FUNCTION  <SNR>33_count()
Called 1 time
Total time:   8.587928
 Self time:   8.587928

count  total (s)   self (s)
    1              0.000003   let l:sum = 0
    1              0.000009   let l:indx = match(a:line, a:pattern)
971648              0.710360   while l:indx >= 0
971648              0.899945     let l:sum += 1
971648              2.405012     let l:match = matchstr(a:line, a:pattern, l:indx)
971648              1.512366     let l:indx += len(l:match)
971648              2.368647     let l:indx = match(a:line, a:pattern, l:indx)
971647              0.444925   endwhile
                              return l:sum

FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    1   8.591591   0.000227  VimtexIndent()
    1   8.590669   0.000064  <SNR>33_indent_delims()
    1   8.587928             <SNR>33_count()
    2   0.002677   0.000652  <SNR>33_split()
    9   0.002025   0.000118  vimtex#util#in_mathzone()
    9   0.001907             vimtex#util#in_syntax()
    2   0.000585   0.000134  <SNR>33_get_prev_line()
    3   0.000476             <SNR>33_is_verbatim()
    3   0.000415             <SNR>14_Highlight_Matching_Pair()
    1   0.000085             <SNR>33_indent_envs()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1              8.587928  <SNR>33_count()
    9              0.001907  vimtex#util#in_syntax()
    2   0.002677   0.000652  <SNR>33_split()
    3              0.000476  <SNR>33_is_verbatim()
    3              0.000415  <SNR>14_Highlight_Matching_Pair()
    1   8.591591   0.000227  VimtexIndent()
    2   0.000585   0.000134  <SNR>33_get_prev_line()
    9   0.002025   0.000118  vimtex#util#in_mathzone()
    1              0.000085  <SNR>33_indent_envs()
    1   8.590669   0.000064  <SNR>33_indent_delims() 
```

**Comment by oliviertilmans** [comment]
Debugging it, both pattern and line in the s:count() function are empty // equal to `''`
s:re_delims is also defined as `['', '', '', '']`

I doubt this is expected, but I am not familiar enough with vimtex/vimL to yet understand why that's the case.

**Comment by lervag** [comment]
No, this is a silly bug from my side. An empty pattern obviously matches everything. I've pushed a quick fix. I don't have time now to test, but I think it should work.

**Comment by oliviertilmans** [comment]
The fix works thanks!

However why can't I get s:re_delims to work in the class file ?
E.g. in the given MWE, if I start in math mode ($ $), I cannot use % to switch between the $'s.
If I rename the file to a *.tex, it works as expected. What should I do to let vimtex know that cls are just plain *.tex files.

**Comment by Rmano** [comment]
@oliviertilmans  observation confirmed. Hang fixed, jumps between math delimiters does not. 
It fixes my example in issue #606, too, so it's clearly a duplicate.

**Comment by lervag** [comment]
The reason that `%` doesn't work is because there is a disruption of the initialization when the main file is not recognized. I can see that this becomes inconvenient in a lot of cases, so I'll fix that. But it will have to be later when I get the time. I've opened a new issue for this to help me remember when I start work on that. See #612.
